### DISCUSSION TITLE
A package with an abstraction layer and extensions for ClickHouse.Client

### DISCUSSION THREAD

**Discussion by MikeAmputer** [DarkWanderer/ClickHouse.Client#524] [Show and tell]
I think the time has come, so i would like to introduce the [ClickHouse.Facades](https://github.com/MikeAmputer/ClickHouse.Facades) package, which is built on `ClickHouse.Client`. It provides an abstraction layer over `ClickHouseConnection` and `ClickHouseCommand`, allowing for better organization and testing of database operations. Additionally, this package offers features like migrations, query cancellation, retries, and more that some may find useful. It is well-documented and production-tested.

The package has some limitations: it is designed specifically for use with dependency injection, supports asynchronous contract only, and targets netstandard2.1. It’s also focused on raw SQL queries only (though, in my opinion, ORMs might not be the best fit for the ClickHouse workflows, and OLAP in general).

I'm curious to hear your thoughts and feedback.

**Comment by DarkWanderer** [comment]
Hello!

This looks like a fairly interesting take on a lightweight ORM. I like the clearly-cut abstractions - e.g. ability to add transactions just by calling a method.

Migrations are something that is widely requested, so having another library covering this is great. I have looked through the code and discovered a lot of features - but having more specific documentation about the specific features (e.g.: what's the best practice to use transactions? how do I combine transactions with retry? etc.) would help it shine

Thank you for sharing this & I am glad that my work helped inspire your creation!
